src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -987,35 +985,51 @@ public string Replace(char oldChar, char newChar)
             }
 
             // Copy the remaining characters, doing the replacement as we go.
-            ref ushort pSrc = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref _firstChar), copyLength);
-            ref ushort pDst = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref result._firstChar), copyLength);
+            ref ushort pSrc = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref _firstChar), (nint)(uint)copyLength);
+            ref ushort pDst = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref result._firstChar), (nint)(uint)copyLength);
+            nint i = 0;
 
             if (Vector.IsHardwareAccelerated && remainingLength >= Vector<ushort>.Count)
             {
-                Vector<ushort> oldChars = new Vector<ushort>(oldChar);
-                Vector<ushort> newChars = new Vector<ushort>(newChar);
+                Vector<ushort> oldChars = new(oldChar);
+                Vector<ushort> newChars = new(newChar);
+
+                Vector<ushort> original;
+                Vector<ushort> equals;
+                Vector<ushort> results;
 
                 do
                 {
-                    Vector<ushort> original = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<ushort, byte>(ref pSrc));
-                    Vector<ushort> equals = Vector.Equals(original, oldChars);
-                    Vector<ushort> results = Vector.ConditionalSelect(equals, newChars, original);
-                    Unsafe.WriteUnaligned(ref Unsafe.As<ushort, byte>(ref pDst), results);
-
-                    pSrc = ref Unsafe.Add(ref pSrc, Vector<ushort>.Count);
-                    pDst = ref Unsafe.Add(ref pDst, Vector<ushort>.Count);
-                    remainingLength -= Vector<ushort>.Count;
+                    original = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<ushort, byte>(ref Unsafe.Add(ref pSrc, i)));
+                    equals = Vector.Equals(original, oldChars);
+                    results = Vector.ConditionalSelect(equals, newChars, original);
+                    Unsafe.WriteUnaligned(ref Unsafe.As<ushort, byte>(ref Unsafe.Add(ref pDst, i)), results);
+
+                    i += Vector<ushort>.Count;
                 }
-                while (remainingLength >= Vector<ushort>.Count);
+                while (i <= (nint)(uint)(remainingLength - Vector<ushort>.Count));
```

Comment:
I suspect this line is not properly hoisted and executed each iteration

Reply:
JIT is smart enough for this (since several versions in my experience).
The loop looks like
```asm
M01_L03:
       lea       r9,[rcx+rcx]
       vmovupd   ymm2,[rax+r9]
       vpcmpeqw  ymm3,ymm2,ymm0
       vpand     ymm4,ymm1,ymm3
       vpandn    ymm2,ymm3,ymm2
       vpor      ymm2,ymm4,ymm2
       vmovupd   [rdx+r9],ymm2
       add       rcx,10
       cmp       rcx,r8
       jle       short M01_L03
```
on x64. Don't know how it looks on ARM, etc. Should I manually introduce a local for this to keep it in a register or is JIT's codepath for the differnt platforms the same here?

Reply:
ah, no it's ok, JIT sometimes doesn't hoist such expressions when it fails to recognize loops as long as it does here it's fine